design-proposals/klip-47-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,122 @@
+* createStream
+* createStreamAs
+* createTable
+* createTableAs
+
+Then inside `AstBuilder#visitCreateStream` and `AstBuilder#visitCreateTable`, the built classes for stream/table 
+instances would be `CreateStream` and `CreateTable`, which are the only two subclasses of `CreateSource`. It makes sense 
+to add the `SOURCE` as a flag into `CreateSource`, and let `AstBuilder` pass the flag into the struct 
+for later stage reference of read-only.
+
+In addition, when user calls `DESCRIBE ... EXTENDED`, the read-only attribute shall be displayed for stream/table. 
+
+### Make Source Stream/Table Read-only
+We plan to add restriction check into the `DistributingExecutor#throwIfInsertOnReadOnlyTopic` or create a similar function, 
+which could verify whether the given stream/table is read-only. If so, the insertion will be rejected.
+
+### Source Table Materialization
+In the call `EngineExecutor#execute`, we will check whether given plan has a `queryPlan` before deciding to proceed 
+creating a persistent query. To bypass this check, we need to provide a concrete query plan within the input plan. We would 
+inject a dummy query plan in `EngineExecutor#plan` phase, when the statement is a CreateTable statement with `SOURCE` tag.
+
+In addition, when user tries to show all the running queries, the materialization query should also be displayed as it takes 
+part of the computation resource. This would be a special type of running query without `SELECT` clause.
+
+Here comes a precaution we might be aware. The original `CREATE TABLE` command does not need materialization, so there is 
+no enforcement for primary key. However, in order to let pull query run on the materialized state, the store must be keyed so that 
+pull query could use the primary key for access.  
+
+To enforce primary key on the `CREATE SOURCE TABLE` command, we have 3 options:
+1. `AstBuilder` could verify that at least one `TableElement` contains primary key, otherwise the plan fails.
+2. `LogicalPlanner#buildPersistentLogicalPlan` could verify the plan node and see whether it needs materialization 
+but has no defined primary key.
+3. Create an injector similar to `TopicCreateInjector`, which could be used to validate things through `RequestValidator`
+
+In long-term, we don't want to bloat `AstBuilder` with all the logical constraint checks. In the short-term, the complexity 
+of #1 is low, and the expected code change to enforce this constraint should be low, so we would pick #1. 
+
+## Public APIS
+* Create source table/stream, which makes the data source read-only:
+
+```roomsql
+CREATE SOURCE STREAM stream_name (COL1 INT, COL2 STRING) AS ...
+CREATE SOURCE TABLE table_name (ID INT KEY, COL1 INT) AS ...
+```
+
+## Test plan
+The new semantics will be tested as:
+
+* Unit testing for modular changes
+* QTTs with the new source table/stream syntax, including insertion failure and successful materialization
+
```

Comment:
We should also have tests for `DESCRIBE` on the sources, and tests to ensure that users are able to drop source streams/tables without a hitch. (Watch out for the case where, today, we disallow dropping sources if there are queries reading from them.)